hubcast/utils/git.py

```diff
@@ -1,17 +1,47 @@
-import os
 import subprocess
 
-# get configuration from environment.
-GIT_REPO_PATH = os.environ.get("HC_GIT_REPO_PATH")
 
+# https://git-scm.com/docs/gitcredentials#_custom_helpers
+def credential_helper(remote: str) -> str:
+    remote = remote.upper()
+    return f"""!f() {{ echo "username=${remote}_USERNAME"; echo "password=${remote}_TOKEN"; }};f"""
 
-def git(args):
-    """Executes a git command on the host system."""
-    result = subprocess.run(
-        ["git", "-C", f"{GIT_REPO_PATH}"] + args.split(),
-        stdout=subprocess.PIPE,
-        stderr=subprocess.STDOUT,
-        text=True,
-        check=True,
-    )
-    return result
+
+class Git:
+    def __init__(
+        self,
+        repo_path: str,
+        gh_username: str,
+        gh_token: str,
+        gl_username: str,
+        gl_token: str,
+        gl_instance_url: str,
+    ):
+        self.repo_path = repo_path
+        self.gl_url = gl_instance_url.removeprefix("https://")
+
+        self.gh_helper = credential_helper("github")
+        self.gl_helper = credential_helper("gitlab")
+
+        self.env = {
+            "GITHUB_USERNAME": gh_username,
+            "GITHUB_TOKEN": gh_token,
+            "GITLAB_USERNAME": gl_username,
+            "GITLAB_TOKEN": gl_token,
+        }
+
+    def __call__(self, cmd: str):
+        """Executes a git command on the host system."""
+        result = subprocess.run(
+            f"git \
+             -c credential.github.com.helper='{self.gh_helper}' \
+             -c credential.{self.gl_url}.helper='{self.gl_helper}' \
+             -C {self.repo_path} {cmd}",
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            env=self.env,
+            shell=True,
```

Comment:
This is a very good catch. I'd debated back and forth about using `shell=True` unfortunately it doesn't appear that you can pass an `env` without running in a shell. We could directly encode the access tokens in the helpers, but we'd be likely to leak them then in logs or error messages from git since they'd be a part of the cmd invocation.

Reply:
Fair enough, we can keep `shell=True`, but then we should probably deploy `shlex.quote` on any args we pass in, to be safe.

Reply:
Okay apparently something was just wrong the first time I tried this. It turns out you can seemingly totally use an `env` dict with no shell. Rewritten and fixed!